Add quick search modal

davidbonan opened the pull request:
**Summary**
Implements a quick search modal for project-wide searching.

**Features**
This modal allows users to quickly search for files and content within the project. It includes a picker for displaying search results and a preview editor for showing the selected result.


https://github.com/user-attachments/assets/ea54a39c-d172-4955-83f3-c50d1d71edea


Performances tested on large codebase (monorepo and zed project).

Related issue https://github.com/zed-industries/zed/issues/8279

Release Notes:

- Added quick search without multibuffer

davidbonan committed:
Adds quick search modal

Implements a quick search modal for project-wide searching.

This modal allows users to quickly search for files and content within the project.
It includes a picker for displaying search results and a preview editor for showing the selected result.

The modal is resizable using drag handles on the right, bottom, and corner.

davidbonan committed:
Improves quick search modal.

Refactors the quick search modal to enhance usability and performance.

It introduces the following changes:

- Persists search queries between openings for a smoother experience.
- Implements file collapsing/expanding for better organization.
- Adds an "Open in Split" button for more flexible workflow.
- Updates the UI, including removing window resizing, to be more efficient and user-friendly, reducing complexity.
- Fixes the preview editor to not allow input.

davidbonan committed:
Adds search options to quick search

Adds case-sensitive, whole-word, regex, and include-ignored
search options to the quick search modal. Also persists the
last search state (query and options) per workspace. Fixes regex
errors and displays in header.

davidbonan committed:
Increases keybinding button size.

Increases the keybinding button size in the quick search picker for better visibility and user experience. The size is adjusted from 11 to 12 rems.

davidbonan commented:
@gustavotrott I have a UX problem with that: 
If I modify a file in the modal, the file is modified but not saved and not opened in the main editor. I find it disturbing to have an unsaved modified file that is not open. Any suggestions?

davidbonan committed:
Improves quick search performance and UX

Refactors quick search to improve performance, especially with
large search results.

Adds constants for max line matches and preview characters, which
enhances configuration and limits resource consumption.

Improves line match navigation.

davidbonan committed:
Improves quick search result extraction.

Refactors quick search to extract file matches and line previews more efficiently.

This change introduces a new function to extract file matches and line previews from a buffer, improving code readability and maintainability. It also leverages SharedString for file keys and parent paths to reduce memory allocations.

Additionally, the quick search modal now accepts an optional initial query, allowing for pre-population of the search field.

davidbonan committed:
Adds click handler to quick search items

Adds a click handler to each quick search item to allow for selecting an item and updating the preview.
Also adds a double click handler to confirm the selected item, dispatching a Confirm action.

davidbonan committed:
Highlights search matches in preview editor.

davidbonan committed:
Focuses the picker on selection

Ensures the quick search picker receives focus when an item is selected.

This improves keyboard navigation and overall user experience
by immediately directing focus to the picker after a selection is made.

davidbonan committed:
Edit and save project directly from preview window

Ensures the editor has access to the project context
when it is created for a buffer, which is needed to
resolve workspace paths.

davidbonan committed:
Removes quick search state persistence

Stops persisting the last quick search state, ensuring a clean state for each new search.

This simplifies the search process by not restoring previous queries and options.

davidbonan committed:
Fixes preview update for non-open files

Ensures the quick search preview updates correctly when the file
to be previewed is not already open in an editor.

davidbonan committed:
Opens previewed file in workspace on edit

Opens the currently previewed file in the workspace when it is edited.

davidbonan committed:
Fix race condition on fast navigation

Ensures that the pending preview path is cleared when the
path is the same as the current one.

This prevents unnecessary reloads when the user is already
viewing the correct content, resolving a race condition
on fast navigation.

davidbonan commented:
Update : 
- Allow modifications directly from the preview, when modifying in the preview, the file opens silently in the main editor.
- Quick Search panel has a fixed size
- Results counter in the header ("42 results in 8 files" or "200+ results (limited)")
- Add search options: Case sensitive, Whole word, Regex, Include ignored
- Collapse/expand files (click on the header)
- Secondary actions: `Cmd+Enter` to open in a split
- When text is selected in the editor, pre-populate the search field with this text when opening the quick search mode.
- Highlight the word in the preview
- Improve UI (square borders, header and footer)
- Improve UX (upon clicking the mouse, we open the preview instead of opening the file directly in the editor and closing the modal)

<img width="1603" height="836" alt="Capture d’écran 2025-12-11 à 10 55 09" src="https://github.com/user-attachments/assets/e8cd1954-edb5-4965-be85-b1f87861e5fb" />
<img width="1603" height="836" alt="Capture d’écran 2025-12-11 à 10 55 24" src="https://github.com/user-attachments/assets/ce215cd4-1458-42c8-8939-2846185fb585" />

davidbonan committed:
Removes unnecessary log dependency

davidbonan committed:
Improves quick search UX and performance

Changes truncate preview to operate on bytes instead of chars, avoiding unicode issues.

Improves performance by caching file matches, and by using binary search to identify visible indices.

Adds a few constants to control timings.

davidbonan committed:
Adds collapse all functionality to quick search

The user can now hold the `Alt` key while clicking a file header to collapse/expand all file headers.

davidbonan committed:
Improves quick search modal responsiveness

Increases the size of the quick search modal and the number of allowed matches.

Ensures the modal adapts to different screen sizes by limiting its dimensions
to 90% of the viewport width and 80% of the viewport height.

davidbonan committed:
Improve tests on Quick Search

baldwindavid commented:
Nice. This is the sort of experience I'd like to see for ALL modals as it's very similar to Helix. I wanted to do something like this for #43961 , but already a stretch to get a new modal. If you get any traction on this, I'd love to build this sort of thing into my PR.

davidbonan commented:
@maxbrunsfeld Thanks for the feedback.

Could you provide one or two concrete examples of what you’re looking for in terms of UX consistency or elegance ?
In particular, if there are specific existing screens, modals, or flows I should align with, that would be very helpful in guiding the changes.

maxbrunsfeld committed:
Prevent row count header from disappearing when searching

maxbrunsfeld commented:
I pushed up one small change to prevent the result count header from briefly disappearing (causing vertical shifts) when searching.

Can you address the latency issue, streaming in the results? Do you have thoughts on how that should work, or if there are any issues with adding that?

baldwindavid commented:
Just off the top of my head some recent PRs, discussions, and issues are related:

- https://github.com/zed-industries/zed/issues/8279#issuecomment-3657220206 - A video of a very similar feature to this one that was not opened as a PR afaik.
- https://github.com/zed-industries/zed/discussions/22581 - Discussion I added earlier this year about leveraging the television TUI for live search of files and content which is similar to this.
- https://github.com/zed-industries/zed/issues/5117 - Issue that @maxbrunsfeld just commented on noting that this sort of UI could be used for find all references.
- https://github.com/zed-industries/zed/pull/44661 - Jump list navigation similar to helix
- https://github.com/zed-industries/zed/pull/43961 - My PR for a netrw / yazi / ranger type file explorer. This attempts to add the sort of functionality from this yazi workaround (https://github.com/zed-industries/zed/discussions/9067) with the existing Zed modal paradigm.

I think that all of these are driving towards a similar request of an experience similar to the one in this quick search PR. This is basically how everything is presented in Helix. Not that we need to copy Helix, but I do think it would be beneficial to take a holistic approach to this. This PR seems like a great one to polish up to set the tone for future pickers.

davidbonan committed:
Improves quick search performance and UX

Add syntax highlighting to the preview.
Caches buffers to improve performance.

maxbrunsfeld commented:
Thanks for explaining; I’ll manually test and review again

maxbrunsfeld commented:
@davidbonan The new UI looks great, streaming is looking good.

I hit a problem when using it in the `chromium` repo, typing in the search bar seemed to freeze the editor. I think there are already some performance problems with regular project search, so I'm trying to figure out if there's anything we need to do in *this PR* to make sure we handle long-running searches properly.

Could you test that scenario when you get a chance? Searching in a large repo like WebKit, Chromium or Linux

justsomelegs commented:
hey @davidbonan nice work!, it might be worth us working together on this tbh, i've been working on basically the same things but with a focus on modular sources #44942.

if not i understand but just a thought?

davidbonan committed:
Adds quick search cancellation support

Implements cancellation support for the quick search feature to prevent unnecessary work when the query changes rapidly.

Shares an atomic boolean between the main thread and worker threads to signal cancellation.

Adds background processing of results to keep the UI responsive.

davidbonan committed:
Improves quick search performance

Reduces UI latency by processing search results in batches,
prioritizing initial results for faster display and preview.

Introduces a cancellation mechanism to prevent processing stale searches, and optimizes text preview generation by safely handling character boundaries.
Also computes search debounce based on project file count to improve responsiveness.

davidbonan committed:
Fix syntax highlighting on buffer insert

Ensures syntax highlighting is updated when a new buffer is added to the delegate's cache.

davidbonan committed:
Fix format

davidbonan committed:
Fix panic

justsomelegs commented:
@davidbonan  When screen width gets too small a stacked view could be cool like the results on top and the preview underneath?

davidbonan committed:
Improves quick search modal sizing

Updates the quick search modal to use dynamic sizing based on viewport dimensions and minimum constraints.

This ensures the modal adapts better to different screen sizes and avoids being too small on larger screens.
It also replaces fixed pixel values with a ratio for the left panel width.

davidbonan committed:
Improves quick search modal layout

Adapts the quick search modal layout to use a vertical
layout when the viewport width is below a certain threshold,
improving usability on smaller screens.

This change also refactors the rendering logic for better
readability and maintainability.

davidbonan committed:
Improves quick search navigation.

Enhances quick search navigation by automatically expanding collapsed file headers when navigating through the list.

davidbonan committed:
Navigates on double click in quick search

This change addresses an issue where double-clicking
an item in quick search wasn't correctly triggering
the navigation. It ensures that the selected index
is properly updated and the confirmation action is
executed, leading to the intended navigation.

davidbonan committed:
Fix clippy

davidbonan committed:
Removes cancellation mechanism

davidbonan committed:
Uses the search results receiver directly

Updates the quick search functionality to directly consume the
search results receiver (`rx`) instead of relying on the `project`
object, streamlining the data flow and improving efficiency.

davidbonan committed:
Yields to UI thread during quick search.

Allows other UI tasks to run between search result batches
to improve responsiveness.

davidbonan committed:
Displays a "Searching..." indicator in Quick Search

Adds a spinner and text label to indicate when Quick Search is actively searching for results. This provides visual feedback to the user that the search is in progress.

davidbonan commented:
> I have landed the project search cancellation changes in https://github.com/zed-industries/zed/pull/45406 (that @maxbrunsfeld paved the way for in https://github.com/zed-industries/zed/pull/45300); I think we're good to remove project search changes from the scope of this PR now.

@osiewicz  Wow, the cancellation mechanism greatly improves the speed of searching in “search on input” mode.
Thx!

wuchunpeng777 commented:
Has this feature been merged? How do I use it?

sf0rman commented:
I really like this, but instead of a popup window, it would be nice if the found files show on the left (where project_panel, git_panel etc go) and the content for each shown in the main screen.

feeiyu commented:
I have a small suggestion after using this feature:
When opening a file from Quick Search and then opening Quick Search again, would it be possible to restore the previous search state (query and results)?

I think this could make repeated navigation a bit smoother.

formbook commented:
thanks for this, was a big feature missing from zed and performance seems excellent. only issue I have with it is it should not automatically search for the symbol under the cursor when launching, thats what gr(get refs) is for. its annoying having to delete the symbol instead of just typing what i want to search for immediately

in neovim with snacks picker, search history can be traversed with ctrl + up/ctrl +down, thats often a pretty useful feature

is there any way to set the font size on the left side results panel?

yara-blue commented:
Hi! Thanks for working on something so requested. There where a few other PR's trying to add a search modal. We picked up another and have now build on it where we feel confident we'll get it merged.

Therefore I am closing this PR. Once we have pickers with preview it will be pretty minimal and only implemented for a few pickers. At that point we'd love your help. If you wanna get notified of the modal landing subscribe to #56037
